api/OBJECT-SIGNING.md

```diff
@@ -0,0 +1,63 @@
+
+## Motivation
+
+The motivation for this is two-fold:
+
+- TLS Man-in-the-Middle proxy products and server-side load balancing and DDoS protection services do not work well with TLS client certificates and the restricted single server root CA certificate used in the V1 API.
+- The desire to be able to carry signed edge-node configuration through some non-TLS channel such as a USB memory stick.
+
+## Certificate management
+
+When TLS MiTM is in place the device will trust the proxy at the TLS level either because it is trusted by being in the standard Linux root certificate list, or because it was specified in the `proxyCertPEM` field in the device configuration API. Thus TLS still provides server authentication and privacy plus integrity.
+
+But since there is no TLS client authentication the `AuthContainer` message is used to sign the payloads from the device to the controller (using standard ECDA signing operations (implemented in ecdsa golang package alternatively the Sign function from github.com/google/go-tpm/tpm2 when the device private key is in a TPM).
+
+The device indicates who it is by providing a hash of its of the certificate corresponding to its signing key in the `AuthContainer`, which is used on the controller side to look up the device (and verify the signature using standard ECDSA verification.)
+Since the device never changes its device certificate per definition (it becomes a new device should it ever change it), there is no need for updating the cert in the device to controller direction.
+
+The logic for the signing and verification is the same in the controller to device direction. However, the certificate handling is different.
+The device will use the `ControllerCerts` API to fetch the signing certificates and any intermediate certificates. The certificate chain will be verified using the standard go X.509's package VerifyChain up to the trusted root CA configured at birth of the device (In EVE this is kept in /config/root-certificate.pem and is orthogonal to the set of trusted root CAs used for TLS).
+The fetch using `ControllerCerts` is required at install time and when the `senderCertHash` does not match a certificate already known to the device. Thus the controller can roll certificates as long as it makes them available using the `ControllerCerts` API and puts the hash of the user signer's certificate in the `senderCertHash` in the `AuthContainer`; that will make the device re-issue a GET of `ControllerCerts`.
+
+During the initial device onboarding the `register` API might be used. Since the controller might not know all of the onboarding certificates a-priori the device MUST include the full `senderCert` in the `AuthContainer`.
+
+## Signing and verification
+
+With the certificates in place the message payloads can be signed and verified on both directions. The 'AuthContainer' has the following fields:
```

Comment:
This is the one section which is missing information, as suggested above. We should explain really clearly how this works, not just the fields. As I understood it, we are looking at something like this, which closely replicates TLS:

1. Sender (device or controller) has a message body to send (probably a protobuf, but it could be gibberish; it doesn't matter)
1. Sender hashes message with known algorithm (specified here)
1. Sender signs hash with sender's private key
1. Sender generates a symmetric encryption key (what algorithm? it can change, but supported ones should be listed here)
1. Sender encrypts entire message - signature and body - with the symmetric key 
1. Sender encrypts symmetric key and its algorithm, e.g. `aes256:abcd15156` with receiver's public key
1. Sender places result as body of message to send: payload, asymmetrically encrypted symmetric key+algo
1. Receiver receives message
1. Receiver decrypts symmetric key+algo; if they cannot, then this is not the receiver
1. Receiver uses decrypted symmetric key to decrypt message
1. Receiver hashes message
1. Receiver uses sender's public key to validate the signature of the hash

This makes the body of the message look like the following:

* `payload` - this is the symmetrically encrypted payload, which contains:
   * original message
   * signed hash of original message
   * algorithm used to sign hash
* `key` - this is the symmetric key and algorithm used to encrypt the payload, plus its algorithm, encrypted with the receiver's public key
* `certificateHash` - this is the hash of the sender's public certificate used to sign, per the original description below

Reply:
There is zero encryption. But I can spell out the steps for the signing and verification using a numbered list.

Reply:
Why not? If our argument is: we lose client TLS auth because of mitm, and all TLS because of non-network channels, so we need to replace those components of TLS; then we should just do it correctly and replace the whole thing? To some degree, it is easier, doing something quite well known.

Reply:
Replacing PFS, replay protection, DoS protection, algorithm negotation in TLS with something that doesn't have a bidirectional channel of communication isn't possible AFAIK. 

Note that the non-network case is only from controller to device and the privacy protection of that information is in the physical protection of the USB stick. However, we do not need full configuration to onboard a device; the most sensitive information are any WiFi credentials on the USB stick. We couldn't even try to encrypt those because the controller does not know the device certificate before it has been onboarded. And it can't onboard until it can talk to the controller.

TLS is a fighter jet. Object signing is a foldable bike loaded in the back seat. Quite useful when arriving at the TLS destination. And a USB stick is just a foldable bike by itself.
No attempt at equivalence between the two what so ever becaus it isn't possible without a communication channel.

Reply:
I like the analogy; made me smile.

I will say that I would be harder-pressed to sell it to a security auditor with signing alone, rather than signing+encryption. Especially because the effort for both is almost identical to the effort for just signing (as listed above).

Reply:
One can only do public key encryption if one knows the receivers cert/public key.
For the yet to be onboarded device needing some network config to be onboarded that isn't possible. And that is the key (only?) current reason we need config on a USB stick. 

The v3 API (which is doing TLS client certs with AuthContainer) is the belts and suspenders approach which works if you don't have MiTM proxies and load balancers on the controller side. It is a lot easier to analyze from a security perspective.